iceoryx_posh/source/runtime/posh_runtime_impl.cpp

```diff
@@ -0,0 +1,579 @@
+// Copyright (c) 2019 - 2021 by Robert Bosch GmbH. All rights reserved.
+// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_posh/internal/runtime/posh_runtime_impl.hpp"
+
+#include "iceoryx_hoofs/cxx/convert.hpp"
+#include "iceoryx_hoofs/cxx/helplets.hpp"
+#include "iceoryx_hoofs/internal/relocatable_pointer/base_relative_pointer.hpp"
+#include "iceoryx_hoofs/posix_wrapper/timer.hpp"
+#include "iceoryx_posh/iceoryx_posh_types.hpp"
+#include "iceoryx_posh/internal/log/posh_logging.hpp"
+#include "iceoryx_posh/internal/runtime/ipc_message.hpp"
+#include "iceoryx_posh/runtime/node.hpp"
+#include "iceoryx_posh/runtime/port_config_info.hpp"
+
+#include <cstdint>
+
+namespace iox
+{
+namespace runtime
+{
+PoshRuntimeImpl::PoshRuntimeImpl(cxx::optional<const RuntimeName_t*> name, const RuntimeLocation location) noexcept
+    : PoshRuntime(name)
+    , m_ipcChannelInterface(roudi::IPC_CHANNEL_ROUDI_NAME, *name.value(), runtime::PROCESS_WAITING_FOR_ROUDI_TIMEOUT)
+    , m_ShmInterface([&] {
+        // in case the runtime is located in the same process like RouDi the shm is already opened;
+        // also in case of the RouDiEnvironment this would close the shm on destruction of the runstime which is also
+        // not desired
+        return location == RuntimeLocation::SAME_PROCESS_LIKE_ROUDI
+                   ? cxx::nullopt
+                   : cxx::optional<SharedMemoryUser>({m_ipcChannelInterface.getShmTopicSize(),
+                                                      m_ipcChannelInterface.getSegmentId(),
+                                                      m_ipcChannelInterface.getSegmentManagerAddressOffset()});
+    }())
+    , m_applicationPort(getMiddlewareApplication())
+{
+}
+
+PoshRuntimeImpl::~PoshRuntimeImpl() noexcept
+{
+    // Inform RouDi that we're shutting down
+    IpcMessage sendBuffer;
+    sendBuffer << IpcMessageTypeToString(IpcMessageType::TERMINATION) << m_appName;
+    IpcMessage receiveBuffer;
+
+    if (m_ipcChannelInterface.sendRequestToRouDi(sendBuffer, receiveBuffer)
+        && (1U == receiveBuffer.getNumberOfElements()))
```

Comment:
Ugs, we need to get rid of these kind of checks :grimacing:

Reply:
Totally agree. We need a proper RPC handling for this